### ISSUE TITLE
VR Healthcheck Fails when a VPC is created, but there are no VMs created in the VPC yet.

### ISSUE THREAD

**Issue by btzq** [apache/cloudstack#10088]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Virtual Router
~~~

##### CLOUDSTACK VERSION

~~~
4.19.1.1
~~~

##### SUMMARY

When a VPC is created, a VR is spawned. 
But, when the VPC does not have any VMs created in it yet, the VR Healthcheck fails with the following error: webserver.service service down at last check 

This gives lots of false alarms for the Cloud Operator because we cant control when the user has created a network or vm or not. The healthcheck fail should only fire if there really is an issue with the router. 

##### STEPS TO REPRODUCE

~~~

~~~

##### ACTUAL RESULTS

~~~
- Create any VPC
- Create any subnet (Eg 1 Tier)
- Do not create any VM
- Monitor alerts, will see VR Healthcheck fail: webserver.service service down at last check 
~~~

##### EXPECTED RESULTS

~~~
- Should not fire healthcheck fail unless there is a real issue.
~~~

**Comment by DaanHoogland** [comment]
@btzq , did you fix a healthcheck script for this?

**Comment by btzq** [comment]
@DaanHoogland what do you mean by that?

**Comment by weizhouapache** [comment]
it looks like "Expected results" and "actual results" should be swapped.

webserver/apache2 (use for metadata/userdata service) is not started if there is no vpc tiers.
it should not be checked in that case.

**Comment by DaanHoogland** [comment]
> webserver/apache2 (use for metadata/userdata service) is not started if there is no vpc tiers. it should not be checked in that case.

yes, sounds like at least one of the check script could be improved. I am trying to figure out which one though. (Maybe it is the starting of the entire run from the java code that should be inhibited)

**Comment by btzq** [comment]
@DaanHoogland no we didnt make any changes to the healthcheck script.

Is there any workaround for us at the moment? We’re getting lots of false alarms.

**Comment by DaanHoogland** [comment]
@btzq , that would involve finding the right check script and adjust it to not cause an alarm, or disable the specific check alltogether.

**Comment by weizhouapache** [comment]
> @btzq , that would involve finding the right check script and adjust it to not cause an alarm, or disable the specific check alltogether.

yes, there is a global setting `network.router.EnableServiceMonitoring` which is default to true
try with disabling it

**Comment by weizhouapache** [comment]
> > > @btzq , that would involve finding the right check script and adjust it to not cause an alarm, or disable the specific check alltogether.
> > 
> > 
> > yes, there is a global setting `network.router.EnableServiceMonitoring` which is default to true try with disabling it
> 
> @weizhouapache will this disable all my healthchecks? I prefer to have the healthchecks, maybe just with the webserver one adjusted.

@btzq 
it will disable the monitoring of services like ssh,dnsmasq, webserver,
you can check the file /etc/monitor.conf in the VRs

ps: I have not tested it yet

**Comment by btzq** [comment]
Ah, that wont do. Actually healthchecks are pretty useful to us, cause usually when it shoots, it really does present something that is wrong. Its only the webserver part thats misleading to us right now. So we cant afford to turn off all health checks.

**Comment by DaanHoogland** [comment]
@btzq , did @weizhouapache answer your queries to satisfaction? (and can we close this issue)

**Comment by btzq** [comment]
@DaanHoogland but the issue hasnt been fixed? 

Refering to the older parts of the conversation, @weizhouapache was mentioning that some improvements on healtcheck script should be improved. 

> it looks like "Expected results" and "actual results" should be swapped.
> 
> webserver/apache2 (use for metadata/userdata service) is not started if there is no vpc tiers. it should not be checked in that case.

**Comment by weizhouapache** [comment]
> [@DaanHoogland](https://github.com/DaanHoogland) but the issue hasnt been fixed?
> 
> Refering to the older parts of the conversation, [@weizhouapache](https://github.com/weizhouapache) was mentioning that some improvements on healtcheck script should be improved.
> 
> > it looks like "Expected results" and "actual results" should be swapped.
> > webserver/apache2 (use for metadata/userdata service) is not started if there is no vpc tiers. it should not be checked in that case.

let me have a quick look

**Comment by weizhouapache** [comment]
@DaanHoogland @btzq 
I was able to reproduce the issue. However, the steps are different from what were described.
- create a VPC
- create a VPC tier. VR health check is ok
- create a VM. VR health check is ok
- destroy the vm, VR health check is ok
- wait until the VPC tier is shutdown, VR health check failed.

`apache2` service in the VPC VR failed with error
```
Mar 07 13:06:10 r-1253-VM apachectl[12046]: AH00558: apache2: Could not reliably determine the server's fully qualified domain name, using 127.0.1.1. Set the 'ServerName' d>
Mar 07 13:06:10 r-1253-VM apachectl[12046]: no listening sockets available, shutting down
Mar 07 13:06:10 r-1253-VM apachectl[12046]: AH00015: Unable to open logs
Mar 07 13:06:10 r-1253-VM apachectl[12043]: Action 'start' failed.
Mar 07 13:06:10 r-1253-VM apachectl[12043]: The Apache error log may have more information.
Mar 07 13:06:10 r-1253-VM systemd[1]: apache2.service: Control process exited, code=exited, status=1/FAILURE
```
this is ok, we should exclude the check on `webserver` and `haproxy` services

**Comment by btzq** [comment]
@weizhouapache so does this mean this will be removed from healthchecks in the next CS Release?

Theres no need to hack scripts or change global settings on my side, right?

**Comment by DaanHoogland** [comment]
> [@weizhouapache](https://github.com/weizhouapache) so does this mean this will be removed from healthchecks in the next CS Release?
> 
> Theres no need to hack scripts or change global settings on my side, right?

No @btzq , we will need to pass a flag to the healthcheck script in situations where these checks need to be skipped. It would not be prudent to skip those unconditionally. Agree @weizhouapache ?

**Comment by weizhouapache** [comment]
> > [@weizhouapache](https://github.com/weizhouapache) so does this mean this will be removed from healthchecks in the next CS Release?
> > Theres no need to hack scripts or change global settings on my side, right?
> 
> No [@btzq](https://github.com/btzq) , we will need to pass a flag to the healthcheck script in situations where these checks need to be skipped. It would not be prudent to skip those unconditionally. Agree [@weizhouapache](https://github.com/weizhouapache) ?

it is a bit complicated, as the services are not monitored by health check scripts.

services are saved at /etc/monitor.conf
in the python script `/root/monitorServices.py`, it always check all services in /etc/monitor.conf
we need to update `/etc/monitor.conf`, or update `/root/monitorServices.py` to be able to exclude the check on some services.

**Comment by weizhouapache** [comment]
> I think adding a parameter to indicate what needs to be excluded seems best. The alternative, editing the conf or script before running, can work as well, but seems more clunky as it then will have to be edited on every add or delete tier, to make sure the right checks are done/skipped. I think it is better to keep the decision as close to the source of truth (orchestration/DB) as possible. 

@DaanHoogland 
the file `/etc/monitor.conf` is created with only ssh when VPC VR starts,
it is updated to include ssh, haproxy, and webserver when VPC tier is added to the VPC VR
however, it is not updated when the only-one VPC tier is shutdown and removed from VPC VR.

the file could be updated when delete a vpc tier, but it looks complicated, and clunky as you said
yes, we need to specify the services need to monitor, based on the guest networks of the VPC VR.

I will work on it

**Comment by weizhouapache** [comment]
fixed by #10710 in ACS 4.22

ssh service is in unknown state, instead of a failure
